src/api-service/__app__/onefuzzlib/azure/nsg.py

```diff
@@ -0,0 +1,311 @@
+            ],
+        )
+
+    if nic.network_security_group and nic.network_security_group.id == nsg.id:
+        return None
+
+    logging.info("associating nic %s with nsg: %s %s", nic.name, resource_group, name)
+
+    nic.network_security_group = nsg
+    network_client = get_network_client()
+    try:
+        network_client.network_interfaces.begin_create_or_update(
+            resource_group, nic.name, nic
+        )
+    except (ResourceNotFoundError, CloudError) as err:
+        if "The request failed due to conflict with a concurrent request" in str(err):
+            logging.debug(
+                "associate NSG with NIC had conflicts",
+                "with concurrent request, ignoring %s",
+                err,
+            )
+            return None
+        return Error(
+            code=ErrorCode.UNABLE_TO_UPDATE,
+            errors=[
+                "Unable to associate nsg %s with nic %s due to %s"
+                % (
+                    name,
+                    nic.name,
+                    err,
+                )
+            ],
+        )
+
+    return None
+
+
+def dissociate_nic(name: str, nic: NetworkInterface) -> Union[None, Error]:
+    if nic.network_security_group is None:
+        return None
+    resource_group = get_base_resource_group()
+    nsg = get_nsg(name)
+    if not nsg:
+        return Error(
+            code=ErrorCode.UNABLE_TO_FIND,
+            errors=["cannot update nsg rules. nsg %s not found" % name],
+        )
+    if nsg.id != nic.network_security_group.id:
+        return Error(
+            code=ErrorCode.UNABLE_TO_UPDATE,
+            errors=[
+                "network interface is not associated with this nsg.",
+                "nsg %s, nic: %s, nic.nsg: %s"
+                % (
+                    nsg.id,
+                    nic.name,
+                    nic.network_security_group.id,
+                ),
+            ],
+        )
+
+    logging.info("dissociating nic %s with nsg: %s %s", nic.name, resource_group, name)
+
+    nic.network_security_group = None
+    network_client = get_network_client()
+    try:
+        network_client.network_interfaces.begin_create_or_update(
+            resource_group, nic.name, nic
+        )
+    except (ResourceNotFoundError, CloudError) as err:
+        if "The request failed due to conflict with a concurrent request" in str(err):
+            logging.debug(
+                "associate NSG with NIC had conflicts with ",
+                "concurrent request, ignoring %s",
+                err,
+            )
+            return None
+        return Error(
+            code=ErrorCode.UNABLE_TO_UPDATE,
+            errors=[
+                "Unable to associate nsg %s with nic %s due to %s"
+                % (
+                    name,
+                    nic.name,
+                    err,
+                )
+            ],
+        )
+
+    return None
+
+
+class NSG(BaseModel):
```

Comment:
This should be in models.py, right? I understand that VM.py has it's own model class defined in the same file, but it seems to me like we should be keeping all the models in models.py

Reply:
I based this on vm.py code. Possible that I should move this to models.py, I'll take a look.

Reply:
My decision point was "If this is intended to be visible to the user in any fashion, then it should go in onefuzztypes. Otherwise, it shouldn't."

Reply:
Broadly scoped, this means anything that was events, requests, and responses go in onefuzztypes.  everything else does not.